crank/tests/test_stack_api.py

```diff
@@ -0,0 +1,192 @@
+"""
+Unit and regression test for the crank package.
+"""
+
+import pytest
+import os, sys, copy, collections
+import numpy as np
+from crank.DihedralScanner import Molecule
+from crank.QMEngine import QMEngine
+from crank.crankAPI import crank_api
+import geometric
+
+try:
+    import psi4
+except:
+    pass
+
+bohr2ang = 0.529177210
+
+class FakeServer:
+    """
+    A fake server, that do these things:
+    1. Take an xyz file and dihedrals list, create json input for crank
+    2. Send the json input dictionary to crankAPI.crank_api(), get the next set of jobs
+    3. Take the set of jobs and run them with engine. Finish if there're no new jobs.
+    4. Collect the results and put them into new json input dictionary
+    5. Go back to step 2.
+    """
+    def __init__(self, xyzfilename, dihedrals, grid_spacing):
+        self.M = Molecule(xyzfilename)
+        self.dihedrals = dihedrals
+        self.grid_spacing = grid_spacing
+        self.engine = Psi4QCEngineEngine()
+        self.engine.M = copy.deepcopy(self.M)
+
+    def run(self):
+        # step 1
+        in_dict = self.create_initial_api_input()
+        while True:
+            # step 2
+            next_jobs = crank_api(in_dict, verbose=True)
+            # step 3
+            if len(next_jobs) > 0:
+                # step 4
+                job_results = self.run_jobs(next_jobs)
+                in_dict = self.update_api_input(in_dict, job_results)
+            else:
+                print("Finished")
+                return self.collect_lowest_energies(in_dict)
+
+    def create_initial_api_input(self):
+        """ Create the initial input dictionary for crank-api """
+        return {
+            'dihedrals': self.dihedrals,
+            'grid_spacing': self.grid_spacing,
+            'elements': self.M.elem,
+            'init_coords': [self.M.xyzs[0].ravel().tolist()],
+            'grid_status': {}
+        }
+
+    def run_jobs(self, next_jobs):
+        """ Take a dictionary of next jobs, run them and return the results """
+        job_results = collections.defaultdict(list)
+        for grid_id_str, job_geo_list in next_jobs.items():
+            grid_id = tuple(int(i) for i in grid_id_str.split(','))
+            dihedral_idx_values = []
+            for dihedral_idxs, dihedral_value in zip(self.dihedrals, grid_id):
+                dihedral_idx_values.append(list(dihedral_idxs) + [dihedral_value])
+            # running jobs in serial here, if run in parallel the order should be kept same
+            for job_geo in job_geo_list:
```

Comment:
There is a lot of technology here, if this much is required it should be moved back inside of Crank itself.

Reply:
I also realized that some knowledge of the dihedral scan process is needed. Since we don't have a universal format for all services to tell what the "server" should do yet, I think there has to be a small module siting on the server, that takes output from crank_api and generate inputs. I agree some of the formatting can be moved into crank_api, but it will make the json data transfer less efficient.

Reply:
The JSON specification is aimed to be a standard. Having utility functions that handle that standard probably isn't a stretch. As it currently stands any changes in Crank have a pretty high chance of breaking the server, I think this approach will be too brittle for scale.